### PR TITLE
2.3 - Add repeat expander feature

### PR DESCRIPTION
**Pull request by LeaklessGfy**
Add "repeat" requested feature in:
https://github.com/coduo/php-matcher/issues/35

And also in:
https://github.com/coduo/php-matcher/issues/21

Feel free to comment, give advice.
I want to merge on 2.3 because I want to stick with PHP 5.6 not 7.

### CODE REVIEW

File: tests/Matcher/Pattern/Expander/RepeatTest.php [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+    {
+        $expander = new Repeat($needle, $isStrict);
+        $this->assertEquals($expectedResult, $expander->match($haystack));
+    }
+
+    public static function examplesProvider()
+    {
+        $jsonPattern = '{"name": "@string@", "activated": "@boolean@"}';
+
+        $jsonTest = array(
```
Comment (eliecharra) [inline tests/Matcher/Pattern/Expander/RepeatTest.php]: You could use "new" PHP short array syntax `[]` which is less verbose

File: src/Matcher/Pattern/Expander/Repeat.php [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+
+        return true;
+    }
+
+    /**
+     * @param array $values
+     * @param Matcher $matcher
+     * @return bool
+     */
+    private function matchJson(array $values, Matcher $matcher)
```
Comment (eliecharra) [inline src/Matcher/Pattern/Expander/Repeat.php]: We should maybe allow this method to be recursive to match nested objects ?

Comment (eliecharra) [reply src/Matcher/Pattern/Expander/Repeat.php]:
```json
{
"offers": "@array@.repeat({\"subscription\": \"@array@.repeat(\"date\": \"@string@.isDateTime()\")\"})"
}
```

This case will work with actual implementation ?

File: src/Matcher/Pattern/Expander/Repeat.php [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+        return self::NAME === $name;
+    }
+
+    /**
+     * @param $value
+     */
+    public function __construct($pattern, $isStrict = true)
+    {
+        if (!is_string($pattern)) {
+            throw new \InvalidArgumentException("Repeat pattern must be a string.");
```
Comment (eliecharra) [inline src/Matcher/Pattern/Expander/Repeat.php]: This exception is not covered by tests

File: src/Matcher/Pattern/Expander/Repeat.php [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+
+    /**
+     * @param array $values
+     * @param Matcher $matcher
+     * @return bool
+     */
+    private function matchScalar(array $values, Matcher $matcher)
+    {
+        foreach ($values as $index => $value) {
+            $match = $matcher->match($value, $this->pattern);
```
Comment (eliecharra) [inline src/Matcher/Pattern/Expander/Repeat.php]: If we only match value, is following pattern failing ?

`@array@.repeat({"foo": "@string@"})`

```json
[
  {
    "foo": "bar" 
  },
  {
    "test": "bar"
  }
]
``

Comment (eliecharra) [reply src/Matcher/Pattern/Expander/Repeat.php]: Ok this is covered by https://github.com/coduo/php-matcher/pull/109/files#diff-a3f73e5a606230d5e7eefa13dddd6221R132

File: README.md [hunk 5]
```diff
@@ -84,6 +84,7 @@ $matcher->getError(); // returns null or error message
 * ``oneOf(...$expanders)`` - example usage ``"@string@.oneOf(contains('foo'), contains('bar'), contains('baz'))"``
 * ``matchRegex($regex)`` - example usage ``"@string@.matchRegex('/^lorem.+/')"``
 * ``optional()`` - work's only with ``ArrayMatcher``, ``JsonMatcher`` and ``XmlMatcher``
+* ``repeat($pattern, $isStrict = true)`` - example usage ``'@array@.repeat({"name" => "foe"})'`` or ``"@array@.repeat('@string@')"``
```
[the remark this answers is not in the document]
Comment (LeaklessGfy) [reply README.md]: Thanks, a little mistake here. Of course, it's JSON. I will fix that ASAP

File: tests/Matcher/Pattern/Expander/RepeatTest.php (hunk 1 above)
Comment (LeaklessGfy) [reply tests/Matcher/Pattern/Expander/RepeatTest.php]: Yes probably, but the whole project seems to use the older `array()` syntax. In order to stick with the project's coding convention I'd rather not change that for now.

File: src/Matcher/Pattern/Expander/Repeat.php (hunk 3 above)
Comment (LeaklessGfy) [reply src/Matcher/Pattern/Expander/Repeat.php]: Will fix that ASAP.

File: src/Matcher/Pattern/Expander/Repeat.php (hunk 4 above)
Comment (LeaklessGfy) [reply src/Matcher/Pattern/Expander/Repeat.php]: I'm not sure to understand what you're talking about here ?

File: src/Matcher/Pattern/Expander/Repeat.php (hunk 2 above)
Comment (LeaklessGfy) [reply src/Matcher/Pattern/Expander/Repeat.php]: This case should work with the nested array.repeat syntax 'cause I just make a call to php-matcher factory. But you're right, maybe allowing nested objects would be more clever.

**Comment by norzechowicz** [comment]
> I want to merge on 2.3 because I want to stick with PHP 5.6 not 7.

Sure, but first I would like to see that feature on master branch and then into 2.3 or even 2.4

File: tests/Matcher/Pattern/Expander/RepeatTest.php (hunk 1 above)
Comment (norzechowicz) [reply tests/Matcher/Pattern/Expander/RepeatTest.php]: On master branch we are using `[]` - 2.3 branch was only created for 5.6 support however I'm not planning to add more new features here without adding them to the master branch first

**Comment by eliecharra** [comment]
> Sure, but first I would like to see that feature on master branch and then into 2.3 or even 2.4

Totally agreed with that

File: src/Matcher/Pattern/Expander/Repeat.php (hunk 4 above)
Comment (eliecharra) [reply src/Matcher/Pattern/Expander/Repeat.php]: Just ignore this comment

**Comment by LeaklessGfy** [comment]
Fix readme and add cover for a new exception in test.

Also create https://github.com/coduo/php-matcher/pull/110 for 3.0 repeat.

File: src/Matcher/Pattern/Expander/Repeat.php (hunk 2 above)
Comment (umpirsky) [reply src/Matcher/Pattern/Expander/Repeat.php]: This is not yet supported, right?

Comment (LeaklessGfy) [reply src/Matcher/Pattern/Expander/Repeat.php]: Yes I'm sorry, didn't have the time to do it

**Pull request merged**